src/main/java/org/spongepowered/common/world/SpongeTeleportHelper.java

```diff
@@ -47,205 +50,131 @@
                 }
+            }
+        }
 
-                // We'll fall onto a block, need to make sure its safe
-                if (typeBelowPos.getType() != BlockTypes.AIR && !isSafeFloorMaterial(BlockUtil.toNative(typeBelowPos).getMaterial())) {
-                    return false;
-                }
+        // Sort them according to the distance to the provided worldLocation.
+        return vectors.stream()
+                .sorted((first, second) -> Integer.compare(vectorLocation.distanceSquared(first), vectorLocation.distanceSquared(second)))
+                .collect(Collectors.toList());
+    }
 
-                // We'll fall through an air block to another, need to make sure
-                // its safe
-                return isSafeFloorMaterial(BlockUtil.toNative(typeBelowPos2).getMaterial());
-            }
+    private BlockData getBlockData(Vector3i vector3i, World world, Map<Vector3i, BlockData> cache) {
+        if (vector3i.getY() < 0) {
+            // Anything below this isn't safe, no point going further.
+            return new BlockData(null);
+        }
 
-            // We have a non-air floor, need to ensure its safe
-            return isSafeFloorMaterial(blockState.getMaterial());
+        if (cache.containsKey(vector3i)) {
+            return cache.get(vector3i);
         }
 
-        // We need to make sure the block at our torso or head is safe
-        return isSafeBodyMaterial(blockState.getMaterial());
+        BlockData data = new BlockData(BlockUtil.toNative(world.getBlock(vector3i)).getMaterial());
+        cache.put(vector3i, data);
+        return data;
     }
 
-    private boolean isSafeFloorMaterial(Material material) {
-        return !(material == Material.CACTUS || material == Material.FIRE || material == Material.LAVA);
-    }
+    private class BlockData {
+
+        private final boolean isSafeBody;
+        private final boolean isSafeFloor;
 
-    private boolean isSafeBodyMaterial(Material material) {
-        return (material == Material.AIR || material == Material.GRASS || material == Material.PLANTS
-                || material == Material.WATER || material == Material.REDSTONE_LIGHT || material == Material.CIRCUITS
-                || material == Material.SNOW || material == Material.PORTAL || material == Material.WEB || material == Material.VINE);
+        // A null material just indicates that we want nothing to be marked as safe.
+        private BlockData(@Nullable Material material) {
+            this.isSafeBody = isSafeBodyMaterial(material);
+            this.isSafeFloor = isSafeFloorMaterial(material);
+        }
+
+        private boolean isSafeFloorMaterial(@Nullable Material material) {
+            return material != null && !(material == Material.AIR || material == Material.CACTUS || material == Material.FIRE
+                    || material == Material.LAVA);
+        }
+
+        private boolean isSafeBodyMaterial(@Nullable Material material) {
```

Comment:
Yeah, went and looked...they do indeed cheat.

As for determining block damage, its a toss-up. Vanilla uses a combination of material checking + collision checks (to prevent spawning in walls). With mods, defining "Does a block hurt me" is incredibly difficult because that definition can change based on the context of the world. @dualspiral 's code works nicely here.